src/properties/propertiesService.ts

```diff
@@ -23,132 +18,71 @@ export interface PropertiesService {
     uiElement: EditorUISchemaElement,
     schemaElement?: SchemaElement
-  ) => {
-    if (
-      schemaElement?.schema.type === 'string' &&
-      !schemaElement?.schema.format &&
-      uiElement.type === 'Control'
-    ) {
-      if (!schemas.schema.properties) {
-        schemas.schema.properties = {};
-      }
-      if (!schemas.schema.properties.options) {
-        schemas.schema.properties.options = {};
-      }
-      assign(schemas.schema.properties.options, {
-        type: 'object',
-        properties: {
-          multi: { type: 'boolean' },
-        },
-      });
-
-      (schemas.uiSchema as Layout).elements.push(
-        createControl('#/properties/options/properties/multi')
-      );
-    }
-    return schemas;
-  },
-};
-
-const labelUIElementDecorator: PropertiesSchemasDecorators = {
-  decorate: (schemas: PropertySchemas, uiElement: EditorUISchemaElement) => {
-    if (uiElement?.type === 'Label') {
-      assign(schemas.schema.properties, { text: { type: 'string' } });
-
-      (schemas.uiSchema as Layout).elements.push(
-        createControl('#/properties/text')
-      );
-    }
-    return schemas;
-  },
-};
-
-const ruleDecorator: PropertiesSchemasDecorators = {
-  decorate: (schemas: PropertySchemas) => {
-    assign(schemas.schema.properties, {
-      rule: {
-        type: 'object',
-      },
-    });
-    (schemas.uiSchema as Layout).elements.push(
-      createControl('#/properties/rule')
-    );
-    return schemas;
-  },
-};
-
-const labelDecorator: PropertiesSchemasDecorators = {
-  decorate: (schemas: PropertySchemas, uiElement: EditorUISchemaElement) => {
-    if (
-      ['Group', 'Control', 'Categorization', 'Category'].includes(
-        uiElement?.type
-      )
-    ) {
-      if (!schemas.schema.properties) {
-        schemas.schema.properties = {};
-      }
-      assign(schemas.schema.properties, { label: { type: 'string' } });
-
-      (schemas.uiSchema as Layout).elements.push(
-        createControl('#/properties/label')
-      );
-    }
-    return schemas;
-  },
-};
-
-const decorators: PropertiesSchemasDecorators[] = [
-  labelDecorator,
-  multilineStringOption,
-  labelUIElementDecorator,
-  ruleDecorator,
-];
+  ) => PropertySchemas;
+}
 export class PropertiesServiceImpl implements PropertiesService {
+  constructor(
+    private schemaProviders: PropertiesSchemasProvider[],
+    private schemaDecorators: PropertiesSchemasDecorator[]
+  ) {}
   getProperties = (
     uiElement: EditorUISchemaElement,
     schemaElement: SchemaElement | undefined
   ): PropertySchemas | undefined => {
-    const decoratedSchemas = decorators.reduce(
+    const providersWithRank = this.schemaProviders.map((provider) => ({
+      provider,
+      rank: provider.tester(uiElement),
+    }));
+    const highestRankedProvider = maxBy(providersWithRank, ({ rank }) => rank);
+    const { provider, rank } = highestRankedProvider || {};
+    if (!provider || !rank || rank === NOT_APPLICABLE) {
```

Comment:
of course you need to adapt this:
```suggestion
   if (provider === undefined || provider.tester(uiElement) === NOT_APPLICABLE) {
```